storage/innobase/log/log0online.cc

```diff
@@ -1132,23 +1268,70 @@ static bool log_online_follow_log_seg(
 
   if (!log_bmp_sys->read_buf.is_data_available()) return true;
 
-  if (log_bmp_sys->has_parse_data_to() >
-      log_bmp_sys->read_buf.get_current_lsn()) {
-    // Seeking to the new read data to start parsing from, next block must be
-    // fully or partially new data to parse
+  ut_ad(log_bmp_sys->read_buf.get_current_lsn() + OS_FILE_LOG_BLOCK_SIZE >
+        log_bmp_sys->start_lsn);
+
+  ulint skip_len{0};
+
+  // If we get here it means all blocks before one containing
+  // log_bmp_sys->start_lsn are already skipped. Now check if we know where
+  // to start parsing the log.
+  if (log_bmp_sys->parse_buf.get_parse_start_lsn() == 0) {
+    // This is the beginning of the iteration for a new checkpoint. At this
+    // point log_bmp_sys->start_lsn contains the previous checkpoint LSN.
+    // As it may point into a middle of a log record we need to find first
+    // record group which starts at or after checkpoint LSN and ignore
+    // everything before this point.
+    auto first_rec_group_offset{0};
+
+    // Looking for the first complete record to start parsing at.
+    while (log_bmp_sys->read_buf.is_data_available() &&
+           (first_rec_group_offset = log_block_get_first_rec_group(
+                log_bmp_sys->read_buf.ccurrent())) == 0) {
+      log_bmp_sys->read_buf.advance();
+    }
```

Comment:
So if I understand correctly, upstream will not use a loop but you use the loop to find the first_rec_group(). Is this is the difference?

Reply:
There is a loop in recovery code as well to iterate all blocks, code is just designed in a different way.